File: C-plus-plus/Binary Search Tree/BinaryTree.cpp
//
// Created by berkan on 11.09.2021.
//

#include "BinaryTree.h"
#include <iostream>

using namespace std;

BinaryTree::BinaryTree()
{
    BinaryTree::root = nullptr;
}

bool BinaryTree::isEmpty()
{
    return BinaryTree::root == nullptr;
}

void BinaryTree::add(Node* node)
{
    if (isEmpty())
    {
        BinaryTree::root = node;
    }
    else
    {
        Node* temp = BinaryTree::root;
        Node* previous = nullptr;

        while (temp != nullptr)
        {
            previous = temp;
            if (node->number > temp->number)
            {
                temp = temp->right;
            }
            else
            {
                temp = temp->left;
            }
        }
        if (node->number > previous->number)
        {
            previous->right = node;
        }
        else
        {
            previous->left = node;
        }
    }
    cout << "Node added" << endl;
}

Node* BinaryTree::maximal(Node* node)
{
    if (isEmpty())
    {
        cout << "Binary Tree is empty" << endl;
        return nullptr;
    }
    else
    {
        Node *temp = node;

        while (temp->right != nullptr)
        {
            temp = temp->right;
        }
        return temp;
    }
}

Node* BinaryTree::minimum(Node *node)
{
    if (isEmpty())
    {
        cout << "Binary Tree is empty" << endl;
        return nullptr;
    }
    else
    {
        Node* temp = node;

        while (temp->left != nullptr)
        {
            temp = temp->left;
        }
        return temp;
    }
}

Node* BinaryTree::del(Node* root, int value)
{
    if (root == nullptr)
    {
        return nullptr;
    }
    else if (root->number == value)
    {
        if (root->left == nullptr && root->right == nullptr)
        {
            delete root;
            return nullptr;
        }
        else if (root->left != nullptr)
        {
            root->number = maximal(root->left)->number;
            root->left = del(root->left, maximal(root->left)->number);
            return root;
        }
        else
        {
            root->number = minimum(root->right)->number;
            root->right = del(root->right, minimum(root->right)->number);
            return root;
        }
    }
    else if (value > root->number)
    {

        root->right = del(root->right, value);
        return root;
    }
    root->left = del(root->left, value);
    return root;
    }


void BinaryTree::show(Node* temp)
{
    if (!isEmpty())
    {
        if (temp->left != nullptr)
        {
            BinaryTree::show(temp->left);
        }
        cout << temp->number << " ";
        if (temp->right != nullptr)
        {
            BinaryTree::show(temp->right);
        }
    }
    else
    {
        cout << "Binary Tree is empty" << endl;
    }
}

File: Python/BinarySearchTree/BinaryTree.py
from node import Node

class BinaryTree:

    def __init__(self):
        self.root:Node = None

    def is_empty(self):
        return self.root == None

    def max_num(self, node:Node):
        if node.right is None:
            return node
        return self.max_num(node.right)

    def min_num(self, node:Node):
        if node.left is None:
            return node
        self.min_num(node.left)

    def add(self, new_node:Node, node:Node, previous:Node = None):
        if self.is_empty():
            self.root = new_node
            return
        if node is None:
            if new_node.num < previous.num:
                previous.left = new_node
            else:
                previous.right = new_node
            print("Node added")
        elif new_node.num < node.num:
            self.add(new_node, node.left, node)
        elif new_node.num > node.num:
            self.add(new_node, node.right, node)

    def preorder(self, node:Node):
        print(node.num, end=" ")
        if node.left is not None:
            self.preorder(node.left)
        if node.right is not None:
            self.preorder(node.right)

    def inorder(self, node:Node):
        if node.left is not None:
            self.inorder(node.left)
        print(node.num, end=" ")
        if node.right is not None:
            self.inorder(node.right)

    def postorder(self, node:Node):
        if node.left is not None:
            self.postorder(node.left)
        if node.right is not None:
            self.postorder(node.right)
        print(node.num, end=" ")

    def delete(self, num, node:Node) -> Node:
        if num == node.num:
            if node.left is None and node.right is None:
                node = None
                return node
            elif node.left is not None:
                node.num = self.max_num(node.left).num
                node.left = self.delete(self.max_num(node.left).num, node.left)
                return node
            else:
                node.num = self.min_num(node.right)
                node.right = self.delete(self.min_num(node.right).num, node.right)
                return node
        elif num < node.num:
            node.left = self.delete(num, node.left)
            return node
        elif num > node.num:
            node.right = self.delete(num, node.right)
            return node